pkg/reconciler/revision/background.go

```diff
@@ -0,0 +1,260 @@
+	for i := 0; i < maxInFlight; i++ {
+		go func() {
+			defer wg.Done()
+			for {
+				item, shutdown := r.queue.Get()
+				if shutdown {
+					return
+				}
+
+				rrItem, ok := item.(*workItem)
+				if !ok {
+					r.logger.Fatalf("Unexpected work item type: want: %T, got: %T", &workItem{}, item)
+				}
+
+				r.processWorkItem(rrItem)
+			}
+		}()
+	}
+
+	// Shut down the queue once the stop channel is closed, this will cause all
+	// the workers to exit.
+	go func() {
+		<-stop
+		r.queue.ShutDown()
+	}()
+
+	// Return a done channel which is closed once all workers exit.
+	done = make(chan struct{})
+	go func() {
+		wg.Wait()
+		close(done)
+	}()
+
+	return done
+}
+
+// Resolve is intended to be called from a reconciler to resolve a revision. If
+// the resolver already has the digest in cache it is returned immediately, if
+// it does not and no resolution is already in flight a resolution is triggered
+// in the background.
+// If this method returns `nil, nil` this implies a resolve was triggered or is
+// already in progress, so the reconciler should exit and wait for the revision
+// to be re-enqueued when the result is ready.
+func (r *backgroundResolver) Resolve(rev *v1.Revision, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) ([]v1.ContainerStatus, error) {
+	r.mu.Lock()
+	defer r.mu.Unlock()
+
+	name := types.NamespacedName{
+		Name:      rev.Name,
+		Namespace: rev.Namespace,
+	}
+
+	result, inFlight := r.results[name]
+	if !inFlight {
+		r.addWorkItems(rev, name, opt, registriesToSkip, timeout)
+		return nil, nil
+	}
+
+	if !result.ready() {
+		return nil, nil
+	}
+
+	ret := r.results[name]
+	return ret.statuses, ret.err
+}
+
+// addWorkItems adds a digest resolve item to the queue for each container in the revision.
+// This is expected to be called with the mutex locked.
+func (r *backgroundResolver) addWorkItems(rev *v1.Revision, name types.NamespacedName, opt k8schain.Options, registriesToSkip sets.String, timeout time.Duration) {
+	r.results[name] = &resolveResult{
+		opt:              opt,
+		registriesToSkip: registriesToSkip,
+		statuses:         make([]v1.ContainerStatus, len(rev.Spec.Containers)),
+		remaining:        len(rev.Spec.Containers),
+		completionCallback: func() {
+			r.enqueue(name)
+		},
+	}
+
+	for i := range rev.Spec.Containers {
+		image := rev.Spec.Containers[i].Image
+
+		r.queue.Add(&workItem{
+			result:  r.results[name],
+			timeout: timeout,
+			name:    rev.Spec.Containers[i].Name,
+			image:   image,
+			index:   i,
+		})
+	}
+}
+
+// processWorkItem runs a single image digest resolution and stores the result
+// in the resolveResult. If this completes the work for the revision, the
+// completionCallback is called.
+func (r *backgroundResolver) processWorkItem(item *workItem) {
+	ctx, cancel := context.WithTimeout(context.Background(), item.timeout)
+	defer cancel()
+
+	result, err := r.resolver.Resolve(ctx, item.image, item.result.opt, item.result.registriesToSkip)
```

Comment:
it is a bit weird that `err` is checked so far from where it is defined.  Perhaps a more unique name to avoid accidental shadowing?

Reply:
good point - done